gitpod-io/gitpod
# [ws-manager] Make cluster selection filter by application cluster

## Description

Redo of #14000 (reverted in https://github.com/gitpod-io/gitpod/pull/14029). The original lacked the changes to `components/gitpod-db/src/typeorm/workspace-cluster-db-impl.ts`. This PR adds that filtering and adds tests for it.

<details>
<summary>Context </summary>
As part of #9198  we want to start syncing the `d_b_workspace_cluster` table with `db-sync`.

Currently, the table differs between US and EU regions because each table contains only the data relevant to that region. For example, in the EU table, the `eu70` workspace cluster is marked as `available` and the `us70` cluster is `cordoned`.  In the US cluster `eu70` is `cordoned` and `us70` is `available`.

In order to sync the table we need to get to a point where there is no difference in the data in the table between EU and US regions.

To do that we will introduce a new field in the table called `applicationCluster` which records the name of the application cluster to which the record belongs. Thus, for each workspace cluster there will be two rows in Gitpod SaaS:

| name | applicationCluster | url | tls | state | ... |
| --- | --- | --- | --- | --- | --- |
| eu70 | eu02 | url | tls info | available | ... |
| eu70 | us02 | url | tls info | cordoned | ... |

Effectively the new `applicationCluster` column gives the table an extra dimension so that we can combine both tables (EU and US) into one.

#13722  added the column to the table and made `gpctl` fill the value when `gpctl register`ing a new workspace cluster. The value is taken from the `GITPOD_INSTALLATION_SHORTNAME` environment variable in `ws-manager-bridge`.
</details>

Make the cluster selection mechanism in `ws-manager-api` aware of the application cluster to which each workspace cluster is registered. Previously, the selection was by workspace cluster name only. Soon, as described in the **context** section, each region in gitpod will store records in the database for *all* workspace clusters - not just the ones in its region. This PR ensures that workspace clusters not in the same region as the application cluster will not be considered.

## Related Issue(s)

Part of https://github.com/gitpod-io/gitpod/issues/9198 and https://github.com/gitpod-io/gitpod/issues/13800

## How to test

Edit the `server` deployment and change the `WSMAN_CFG_MANAGERS` environment variable to change the `applicationCluster` to which the default workspace cluster in the preview environment is registered:

```json
[
  {
    "name": "",
    "url": "dns:///ws-manager:8080",
    "tls": {
      "ca": "/ws-manager-client-tls-certs/ca.crt",
      "crt": "/ws-manager-client-tls-certs/tls.crt",
      "key": "/ws-manager-client-tls-certs/tls.key"
    },
    "state": "available",
    "maxScore": 100,
    "score": 50,
    "govern": true,
    "admissionConstraints": null,
    "applicationCluster": "dev" // <- change this from "dev" to anything else
  }
]
```
The `WSMAN_CFG_MANAGERS` environment variable is base64 encoded.

Once the `server` deployment is edited it should no longer be possible to start a workspace as the only workspace cluster is now associated with a different (non-existent) application cluster:

<img width="743" alt="image" src="https://user-images.githubusercontent.com/8225907/197509275-61ebccb0-dace-44d0-9b54-72fbe3cb4629.png">

Edit the `WSMAN_CFG_MANAGERS` environment variable once more and set the `applicationCluster` field back to `"dev"`. Starting a workspace should now proceed as normal.

## Release Notes

```release-note
NONE
```

## Documentation


## Werft options:

      If enabled this will build `install/preview`
      Valid options are `all`, `workspace`, `webapp`, `ide`
